persistence/data_models_json_fields.go

```diff
@@ -91,6 +92,127 @@ func (s *StateExecutionSequenceMapsJson) CompleteNewStateExecution(stateId strin
 	return nil
 }
 
+type StateExecutionWaitingQueuesJson struct {
+	// { queue_1: {(state_1, state_1_seq): waiting_count, (state_2, state_2_seq): waiting_count, ...} }
+	QueueToStatesMap map[string]map[string]int `json:"queueToStatesMap"`
+	// { (state_1, state_1_seq): total_waiting_count, (state_2, state_2_seq): total_waiting_count, ...} }
+	StateToQueueCountMap map[string]int `json:"stateToQueueCountMap"`
+	// [ message_1, message_2, ... ]
+	UnconsumedMessages []LocalQueueMessageInfoJson `json:"unconsumedMessages"`
+}
+
+func NewStateExecutionWaitingQueues() StateExecutionWaitingQueuesJson {
+	return StateExecutionWaitingQueuesJson{
+		QueueToStatesMap:     map[string]map[string]int{},
+		StateToQueueCountMap: map[string]int{},
+		UnconsumedMessages:   []LocalQueueMessageInfoJson{},
+	}
+}
+
+func (s *StateExecutionWaitingQueuesJson) ToBytes() ([]byte, error) {
+	return json.Marshal(s)
+}
+
+func NewStateExecutionWaitingQueuesFromBytes(bytes []byte) (StateExecutionWaitingQueuesJson, error) {
+	var waitingQueuesJson StateExecutionWaitingQueuesJson
+	err := json.Unmarshal(bytes, &waitingQueuesJson)
+	return waitingQueuesJson, err
+}
+
+func (s *StateExecutionWaitingQueuesJson) AddNewLocalQueueCommandForStateExecution(
+	stateExecutionId StateExecutionId, command xdbapi.LocalQueueCommand, anyOfCompletion bool) {
+	count := command.GetCount()
+	if count == 0 {
+		count = 1
+	}
+
+	stateExecutionIdKey := stateExecutionId.GetStateExecutionId()
+
+	m, ok := s.QueueToStatesMap[command.GetQueueName()]
+	if !ok {
+		m = map[string]int{}
+	}
+
+	if anyOfCompletion {
+		m[stateExecutionIdKey] = 1
+	} else {
+		m[stateExecutionIdKey] += int(count)
+	}
+	s.QueueToStatesMap[command.GetQueueName()] = m
+
+	if anyOfCompletion {
+		s.StateToQueueCountMap[stateExecutionIdKey] = 1
+	} else {
+		s.StateToQueueCountMap[stateExecutionIdKey] += int(count)
+	}
+}
+
+// Consume return assigned StateExecutionId string and a boolean indicating whether the assigned StateExecutionId has finished waiting
+func (s *StateExecutionWaitingQueuesJson) Consume(message LocalQueueMessageInfoJson) (*string, bool) {
+	m, ok := s.QueueToStatesMap[message.QueueName]
+	if !ok {
+		// no states have started waiting for this queue
+		s.UnconsumedMessages = append(s.UnconsumedMessages, message)
+		return nil, false
+	}
+
+	keys := make([]string, 0, len(m))
+	for key := range m {
+		keys = append(keys, key)
+	}
+
+	// E.g., given m as:
```

Comment:
Thanks for having those comments. It is extremely helpful for me to understand. 

I think we are missing something here -- should a LocalQueueCommand(with count>1) consume a message even the message is not enough to fulfill the whole command? 

Consider this example:
stateExeA waits for q1 for 2 msg, and a message comes, so it's assigned to stateExeA but it can't complete(needing one more).

But then stateExeB waits for q1 for 1 msg, should we move the 1 msg to B so it it can complete? 

This is related to what we discussed early for "QueueTypes":
* FCFS type: whoever waits for the queue name will consume the messages for the queue first
* Best first match for command: we won't assign the message unless the whole localQueueCommand can be fulfilled 

Looks this implementation is between the two Lol(very creative btw, I never thought about this idea) -- it's sorting based on current waiting count, and always assign to whoever wait the least. It's like a heuristic optimization. It will works but I feel like it will not be easily understood. 

Based on that thinking, I feel like we probably can change the `QueueToStatesMap` / `StateToQueueCountMap` to just store stateExe to their localQueue commands directly (because merging them with queue names will lose some info, for example, a stateExe could have multiple localQueue commands for the same queueName). And change `UnconsumedMessages` to group by the queue name. When a new messages arrive, just append/add the msg to the UnconsumedMessages and try consume based on the whole localQCommand.